### ISSUE TITLE
flash doesn't contain any data

### ISSUE THREAD

**Issue by Logosoft** [playframework/playframework#7533]
### Are you looking for help?

No

### Play Version (2.5.x / etc)

2.6.0

### API (Scala / Java / Neither / Both)

Scala

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Windows 8.1

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) Client VM (build 25.131-b11, mixed mode, sharing)

### Library Dependencies

No dependencies

### Expected Behavior

`request.flash` should contain data passed using such code
`Home.flashing("success" -> "Computer has been created")`

It worked fine with Play 2.5.x in particular in the Anorm example project
https://example.lightbend.com/v1/download/play-scala-anorm-example-2.5.x

How to check:
1. Run project and apply database evolution.
2. Click on "ACE" computer name.
3. Click on "Delete this Computer"

Result:
![good](https://user-images.githubusercontent.com/12233553/27699024-989daf1e-5d01-11e7-99b1-44014dedcc0d.png)

### Actual Behavior

The same steps with Anorm example project for Play 2.6.0 (https://example.lightbend.com/v1/download/play-scala-anorm-example) produce such result:

![bad](https://user-images.githubusercontent.com/12233553/27699134-f3530d64-5d01-11e7-9ec0-882867173605.png)

The issue is also appears in case of creating seed project using `sbt new playframework/play-scala-seed.g8` and in a controller adding something like `Home.flashing("success" -> "Computer has been created")`. In views the code `request.flash.data.size` always returns 0.

### Reproducible Test Case

Test cases for both Play 2.5.x and 2.6.0 pass correctly. The problem only appears when the project is tested manually in browser.

**Comment by richdougherty** [comment]
Starting with `play-scala-seed` I wasn't able to reproduce:

Routes:
```
GET     /                           controllers.HomeController.index
GET     /redirected                           controllers.HomeController.redirected
```

HomeController:
```scala
  def index() = Action { implicit request: Request[AnyContent] =>
    Redirect("/redirected").flashing("success" -> "Computer has been created")
  }

  def redirected() = Action { implicit request: Request[AnyContent] =>
    Ok(request.flash.toString)
  }
```

Network headers seen in Chrome 59 on Mac OS 10.12.5:

Response to `http://localhost:9000/`:

```
HTTP/1.1 303 See Other
Location: /redirected
Set-Cookie: PLAY_FLASH=eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7InN1Y2Nlc3MiOiJDb21wdXRlciBoYXMgYmVlbiBjcmVhdGVkIn0sIm5iZiI6MTQ5OTAyNTM3OSwiaWF0IjoxNDk5MDI1Mzc5fQ.rFPs1p6w432kVSabiSpLaQz4obzM0pKWwiw_oqPs1PQ; SameSite=Lax; Path=/; HTTPOnly
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Permitted-Cross-Domain-Policies: master-only
Date: Sun, 02 Jul 2017 19:56:19 GMT
Content-Length: 0
```

Response to `http://localhost:9000/redirected`:
```
HTTP/1.1 200 OK
Set-Cookie: PLAY_FLASH=; Max-Age=-86400; Expires=Sat, 01 Jul 2017 19:56:19 GMT; Path=/; HTTPOnly
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Permitted-Cross-Domain-Policies: master-only
Date: Sun, 02 Jul 2017 19:56:19 GMT
Content-Type: text/plain; charset=utf-8
Content-Length: 48

Flash(Map(success -> Computer has been created))
```

Can you please confirm with your browser version info and show the headers you get back for the redirect and the view?

**Comment by Logosoft** [comment]
I have tested everything again.

1. Your example works fine.

2. I've compared Anorm Example project for versions 2.5 and 2.6. In header25.txt and header26.txt you can see headers received after deleting a computer as was described earlier. In case of version 2.6 it doesn't work as should and also it generates warning during runtime (see warning.txt).

2. I prepared an example from seed project and used there patterns similar to Anorm Example project. Please, see PlayTest.zip. If you run it and go to http://localhost:9000/delete, you will be redirected to http://localhost:9000/redirected and will see flash with correct data. But if you press "Click me" button on http://localhost:9000/redirected or on http://localhost:9000/ you will see empty flash though the button activates the same action as http://localhost:9000/delete. The difference is only in GET or POST method used.
There are some headers: in click.txt header received by clicking the button, in delete.txt header related to http://localhost:9000/delete and in redirected.txt header after redirect. Headers were received in Chrome Version 58.0.3029.110 (64-bit) in Windows 8.1.
[PlayTest.zip](https://github.com/playframework/playframework/files/1118150/PlayTest.zip)
[header25.txt](https://github.com/playframework/playframework/files/1118154/header25.txt)
[header26.txt](https://github.com/playframework/playframework/files/1118152/header26.txt)
[redirected.txt](https://github.com/playframework/playframework/files/1118151/redirected.txt)
[warning.txt](https://github.com/playframework/playframework/files/1118156/warning.txt)
[click.txt](https://github.com/playframework/playframework/files/1118153/click.txt)
[delete.txt](https://github.com/playframework/playframework/files/1118155/delete.txt)

**Comment by richdougherty** [comment]
Thanks for that info. I have reproduced now on 2.6.0 using the Anorm example template you linked above. I get the same error in the console. It looks like Play's doesn't trust the Flash cookie value the browser provides in the `Cookie` header. It's possible that there's some config migration needed in the Anorm example template, but I'm not familiar with this area of Play so I'm not sure.

```
[warn] p.a.m.DefaultJWTCookieDataCodec - decode: could not decode JWT: JWT signature does not match locally computed signature. JWT validity cannot be asserted and should not be trusted.
io.jsonwebtoken.SignatureException: JWT signature does not match locally computed signature. JWT validity cannot be asserted and should not be trusted.
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:354)
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:481)
	at io.jsonwebtoken.impl.DefaultJwtParser.parseClaimsJws(DefaultJwtParser.java:541)
	at play.api.mvc.JWTCookieDataCodec$JWTFormatter.parse(Cookie.scala:652)
	at play.api.mvc.JWTCookieDataCodec.decode(Cookie.scala:577)
	at play.api.mvc.JWTCookieDataCodec.decode$(Cookie.scala:569)
	at play.api.mvc.DefaultJWTCookieDataCodec.decode(Cookie.scala:742)
	at play.api.mvc.FallbackCookieDataCodec.decode(Cookie.scala:728)
	at play.api.mvc.FallbackCookieDataCodec.decode$(Cookie.scala:719)
	at play.api.mvc.DefaultFlashCookieBaker.decode(Flash.scala:91)
```

The relevant headers are:

```
POST /computers/381/delete HTTP/1.1
Cookie: PLAY_SESSION=eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7ImNzcmZUb2tlbiI6IjM1MmEwMzgyY2FkMzc5MDM5MzcxMzlmYTRmYWJkNDE3NDQzMzg4NzAtMTQ5OTA2NjM0NTkxMC1lMDAwN2U5YTlkMjE2MjA4YmMzZTdlZmEifSwibmJmIjoxNDk5MDY2MzQ1LCJpYXQiOjE0OTkwNjYzNDV9.ZixAKHxhtg5BMSwZF5Lzy9wXU2aSQgkD-AX8h4TZSNY

HTTP/1.1 303 See Other
Location: /computers
Set-Cookie: PLAY_FLASH=eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7InN1Y2Nlc3MiOiJDb21wdXRlciBoYXMgYmVlbiBkZWxldGVkIn0sIm5iZiI6MTQ5OTA3MDAyMCwiaWF0IjoxNDk5MDcwMDIwfQ.PofrruigaE3IMZ9Zkyi0ZAynja4ctiB7hUBz2BicixY; SameSite=Lax; Path=/; HTTPOnly

GET /computers HTTP/1.1
Cookie: PLAY_SESSION=eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7ImNzcmZUb2tlbiI6IjM1MmEwMzgyY2FkMzc5MDM5MzcxMzlmYTRmYWJkNDE3NDQzMzg4NzAtMTQ5OTA2NjM0NTkxMC1lMDAwN2U5YTlkMjE2MjA4YmMzZTdlZmEifSwibmJmIjoxNDk5MDY2MzQ1LCJpYXQiOjE0OTkwNjYzNDV9.ZixAKHxhtg5BMSwZF5Lzy9wXU2aSQgkD-AX8h4TZSNY; PLAY_FLASH=eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7InN1Y2Nlc3MiOiJDb21wdXRlciBoYXMgYmVlbiBkZWxldGVkIn0sIm5iZiI6MTQ5OTA3MDAyMCwiaWF0IjoxNDk5MDcwMDIwfQ.PofrruigaE3IMZ9Zkyi0ZAynja4ctiB7hUBz2BicixY

HTTP/1.1 200 OK
...
```

**Comment by Logosoft** [comment]
It seems the issue is somehow related to database evolutions. In case of any evolution required `request.flash` will be empty. It means that when in-memory database is used flash will be always empty. When database is located in a disc storage, it is necessary to restart an application after applying an evolution to make flash working.
The issue doesn't depend on further database access.

How to test using  [Play Scala Anorm Example](https://example.lightbend.com/v1/download/play-scala-anorm-example):

1. Run project and apply database evolution.
2. Click on "ACE" computer name.
3. Click on "Delete this Computer". Message about computer deletion doesn't appear.

How make it working:

1. Reconfigurate database in application.conf. For example comment this row `db.default.url="jdbc:h2:mem:play"` and add `db.default.url="jdbc:h2:~/sample"`
2. Run project and apply database evolution.
3. Click on "ACE" computer name.
4. Click on "Delete this Computer". Message about computer deletion doesn't appear.
5. Restart the application.
6. Click on "Acer Extensa" computer name.
7. Click on "Delete this Computer". Message about computer deletion appears. So flash is correct now.

Steps 3-4 are optional just to make sure that flash is empty after applying evolution.

Another way to test is to use [Temp.zip](https://github.com/playframework/playframework/files/1118899/Temp.zip) project, which is created from Play Scala Anorm Example. Here dummy data are used and there is no reading from database. So the project can be tested without evolutions. Steps for testing are the same.

**Comment by neVERberleRfellerER** [comment]
For p.a.m.DefaultJWTCookieDataCodec error, try "-Duser.timezone=Europe/London". Not sure why this happens, but I am facing this problem: sessions don't work at all. With timezone set to Europe/London, everything works.

**Comment by marcospereira** [comment]
This:

```
[warn] p.a.m.DefaultJWTCookieDataCodec - decode: could not decode JWT: JWT signature does not match locally computed signature. JWT validity cannot be asserted and should not be trusted.
io.jsonwebtoken.SignatureException: JWT signature does not match locally computed signature. JWT validity cannot be asserted and should not be trusted.
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:354)
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:481)
	at io.jsonwebtoken.impl.DefaultJwtParser.parseClaimsJws(DefaultJwtParser.java:541)
	at play.api.mvc.JWTCookieDataCodec$JWTFormatter.parse(Cookie.scala:652)
	at play.api.mvc.JWTCookieDataCodec.decode(Cookie.scala:577)
	at play.api.mvc.JWTCookieDataCodec.decode$(Cookie.scala:569)
	at play.api.mvc.DefaultJWTCookieDataCodec.decode(Cookie.scala:742)
	at play.api.mvc.FallbackCookieDataCodec.decode(Cookie.scala:728)
	at play.api.mvc.FallbackCookieDataCodec.decode$(Cookie.scala:719)
	at play.api.mvc.DefaultFlashCookieBaker.decode(Flash.scala:91)
```

Is related to #7530. If you try using a private/incognito tab where there are not previous cookies, then you shouldn't see this warning.

**Comment by marcospereira** [comment]
Sorry, it is not related to the issue above.

I was able to reproduce it too.

**Comment by richdougherty** [comment]
> code in #7533 (comment) does NOT work without timezone set to Europe/London

I'm pretty sure my timezone is `Pacific/Auckland` and it worked for me.

However, if changing the timezone changes the behaviour for you then that's a good clue. It could be related to `Cookie` expiration or JWT might have a lease time on its tokens.

**Comment by richdougherty** [comment]
I think we need to put a breakpoint into https://github.com/jwtk/jjwt/blob/master/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java#L351 to see what is invalid about the signature.

**Comment by richdougherty** [comment]
I'm going to take a shot at debugging this now.

**Comment by richdougherty** [comment]
I've found the problem and started working on an initial pull request: #7556.

@Logosoft, thanks for reporting this—it was a tricky one!

@neVERberleRfellerER, if you can reproduce a timezone bug, would you mind creating another issue that describes the problem in detail. Thanks!

**Comment by neVERberleRfellerER** [comment]
@richdougherty There probably isn't separate bug, just failure on my side. When I was reproducing this error with your code (https://github.com/playframework/playframework/issues/7533#issuecomment-312513723), I created project from seed, ran it, modifying code while it was starting -- since it takes a while -- which caused failure (since in both cases I have changed routes first and they pointed to non existent action in controller) and broke the keys. Then it worked with timezone, because it was different clean start. Other projects were broken by evolutions. My own test mini project without evolutions was borked by unbealiaveble cascade that contained every possible wrong step multiplied by https://github.com/playframework/playframework/pull/7530 (since cookies were signed by invalid key). Your test code indeed works even without any timezone changes after I started it again with clean browser.
I apologize for my derailing analysis.
